### PR TITLE
Add REST for LA-PHL and Refactor

### PR DESCRIPTION
**Pull request by oslynn**
This PR added REST transport support for LA-PHL and refactored it.  I illuminated the hardcode dependencies to URL endpoints of STLTs.  Now, it uses the information from the transport settings to determine the authentication and send reports to STLTs.

Test Steps:
1. ./gradlew clean package
2. docker-compose up
3. Populate the local vault with credentials for the following STLTs:
    DATAINGESTION--DATATEAM-CDC-NBS,  FLEXION--ETOR-SERVICE-RECEIVER, LA-PHL--ETOR-NBS-ORDERS, OK-PHD--ELR, NY-PHD--ELR, NY-PHD--FULL-ELR-TEST, WA-PHD--REST.
4. Upload settings
    ./prime multiple-settings set --input ./settings/staging/0160-nbs-staging-new-rest-setting.yml
   ./prime multiple-settings set --input ./settings/staging/0161-flexion-staging-new-rest-setting.yml
  ./prime multiple-settings set --input ./settings/staging/0162-la-phl-staging-new-rest-setting.yml
  ./prime multiple-settings set --input ./settings/staging/0163-ok-phd-staging-new-rest-setting.yml
  ./prime multiple-settings set --input ./settings/staging/0164-ny-staging-new-rest-setting.yml
5. Run STLT's check as following:
  curl "http://localhost:7071/api/check?sftpcheck=dataingestion.datateam-cdc-nbs"
  Result:
  image
 curl "http://localhost:7071/api/check?sftpcheck=flexion.etor-service-receiver"
 Result:
image
curl "http://localhost:7071/api/check?sftpcheck=la-phl.etor-nbs-orders"
Result:
image
curl "http://localhost:7071/api/check?sftpcheck=ok-phd.elr"
Result:
image
curl "http://localhost:7071/api/check?sftpcheck=ny-phd.elr"
Result:
image
curl "http://localhost:7071/api/check?sftpcheck=ny-phd.full-elr-test"
Result:
image
curl "http://localhost:7071/api/check?sftpcheck=wa-phd.rest"
Result:
image

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

### Testing
- [x] Tested locally?
- [x] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [x] Ran ./gradlew testIntegration
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/settings/prod/0160-ok-phd-prod-new-rest-setting.yml [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# To submit this to localhost, run:
+#
+# ./prime multiple-settings set --env prod --input ./settings/prod/0160-ok-phd-prod-new-rest-setting.yml
```
Comment (adegolier) [inline prime-router/settings/prod/0160-ok-phd-prod-new-rest-setting.yml:3]: This doesn't submit to localhost

File: prime-router/settings/prod/0161-wa-prod-new-rest-setting.yml [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# To submit this to localhost, run:
+#
+# ./prime multiple-settings set --env prod --input ./settings/prod/0161-wa-prod-new-rest-setting.yml
```
Comment (adegolier) [inline prime-router/settings/prod/0161-wa-prod-new-rest-setting.yml:3]: Doesn't submit to localhost

File: prime-router/settings/prod/0162-ny-prod-new-rest-setting.yml [hunk 3]
```diff
@@ -0,0 +1,279 @@
+# To submit this to localhost, run:
+#
+# ./prime multiple-settings set --env prod --input ./settings/prod/0162-ny-prod-new-rest-setting.yml
```
Comment (adegolier) [inline prime-router/settings/prod/0162-ny-prod-new-rest-setting.yml:3]: Doesn't submit to localhost

File: prime-router/settings/staging/0162-la-phl-staging-new-rest-setting.yml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        onlyOncePerDay: false
+    description: ""
+    transport: !<REST>
+      reportUrl: "https://api.neometrics.com/natusAPI/api/v2/HL7"
+      authTokenUrl: "https://api.neometrics.com/natusAPI/api/v2/AUTH/Login"
+      headers:
+        Content-Length:  "<calculated when request is sent>"
+        Content-Type:  "multipart/form-data"
+        Key: "files"
+        Subscription: "23edf66e1fe14685bb9dfa2cbb14eb3b"
```
Comment (adegolier) [inline prime-router/settings/staging/0162-la-phl-staging-new-rest-setting.yml:93]: Is.... it okay for this to be in here?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 5]
```diff
@@ -94,11 +94,6 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
         // get the username/password to authenticate with OAuth
         val (credential, jksCredential) = getCredential(restTransportInfo, receiver)
 
-//         restTransportInfo.headers = restTransportInfo.headers + Pair("Content-Length", header.content.size.toLong())
-
-        // get the TLS/SSL cert in a JKS if needed, NY uses a specific one
-//        val jksCredential = restTransportInfo.tlsKeystore?.let { lookupJksCredentials(it) }
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:100]: 🎉 ty for removing commented out code!

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 6]
```diff
@@ -413,36 +408,72 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+            val idTokenInfoString: String = client.post(restUrl) {
+                expectSuccess = true // throw an exception if not successful
+                postHeaders(
+                    if (restTransportInfo.authHeaders["Authorization-Type"] == "Basic Auth") {
+                        // Authorization-Type: "Basic Auth" requires the following:
+                        // Header:
+                        //  Authorization : "Basic + base64(username+password)"
+                        // Body:
+                        //  None
+                        // TODO: Add Authorization-Type: "Basic Auth" to DATAINGESTION (NBS) RESTTransport.authHeaders
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:427]: I noticed that "Basic Auth" is in the NBS Staging settings.... does this need to be done for prod, too, or is this referring to something different?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 7]
```diff
@@ -413,36 +408,72 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+                    setBody(mapOf("username" to credential.user, "password" to credential.pass))
+                } else if (restTransportInfo.authHeaders["Authorization-Type"] == "email/password") {
+                        // Authorization-Type: email/password requires the following:
+                // Header:
+                //  Content-Type: application/json
+                //  Authorization: username/password
+                // Body:
+                // { "email": "<email@domain.com>", "password": "<password>"
+                    // OK-PHD.
+                    // TODO: Add Authorization-Type: "email/password" to OK-PHD RESTTransport.authHeaders
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:462]: This seems to be done for both staging and production settings.... is this referring to something else or can this TODO be removed?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 8]
```diff
@@ -470,41 +501,44 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+                    }.toMap()
                 )
+
                 setBody(
-                    when (restUrl.substringAfterLast('/')) {
-                        // OK or NBS
-                        "hl7", "reports" -> {
+                    when (headers["Content-Type"]) {
+                        // OK or NBS Content-Type: text/plain
+                        // TODO: Add Content-Type: text/plain to OK-PHD, NBS (DATAINGESTION) payload header of RESTTransport
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:519]: This seems to have been done in the settings?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 9]
```diff
@@ -470,41 +501,44 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+                        "text/plain" -> {
                             TextContent(message.toString(Charsets.UTF_8), ContentType.Text.Plain)
                         }
-                        // Flexion
-                        "demographics" -> {
-                            TextContent(message.toString(Charsets.UTF_8), ContentType.Application.Json)
-                        }
-                        "orders" -> {
+                        // Flexion Content-Type: text/fhir+ndjson
+                        // TODO: Add Content-Type: text/fhir+ndjson to Flexion.etor-service-receiver payload header of RESTTransport
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:524]: Seems to have been done in settings?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 10]
```diff
@@ -470,41 +501,44 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+                        "text/fhir+ndjson" -> {
                             TextContent(message.toString(Charsets.UTF_8), ContentType.Application.Json)
                         }
-                        // WA
-                        "elr" -> {
+                        // WA Content-Type: application/json
+                        // TODO: Add Content-Type: application/json to WA-PHD payload header of RESTTransport
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:529]: Seems to have been done in settings?

File: prime-router/src/main/kotlin/transport/RESTTransport.kt [hunk 11]
```diff
@@ -470,41 +501,44 @@ class RESTTransport(private val httpClient: HttpClient? = null) : ITransport {
...
+                        "application/json" -> {
                             contentType(ContentType.Application.Json)
                             // create JSON object for the BODY. This encodes "/" character as "//", needed for WA to accept as valid JSON
                             JSONObject().put("body", message.toString(Charsets.UTF_8)).toString()
                         }
-                        else -> {
-                            // NY
+                        // NY Content-Type: multipart/form-data
+                        // TODO: Add Content-Type: multipart/form-data to NY-PHD.ELR, NY-PHD.FULL-ELR-TEST, LA-PHL (Natus) payload header of RESTTransport
```
Comment (adegolier) [inline prime-router/src/main/kotlin/transport/RESTTransport.kt:536]: Seems to have been done in settings?

**Comment by adegolier** [comment]
Followed test instructions, OK is the only one that succeeded

**Review by victor-chaparro — COMMENTED** [review]
Is there anyway we can actually see what the differences in these files are?

@victor-chaparro, the only differences that show are in RESTTransport.kt.  All setting files, we always create new files when we update the setting.

**Review by victor-chaparro — APPROVED** [review]
Good to see a lot of hard coded logic is removed

**Pull request merged**